### PR TITLE
[SPARK-23912][SQL][Followup] Refactor ArrayDistinct

### PR DESCRIPTION
**Pull request by kiszk**
## What changes were proposed in this pull request?

This PR simplified code generation for `ArrayDistinct`. #21966 enabled code generation only if the type can be specialized by the hash set. This PR follows this strategy.

Optimization of null handling will be implemented in #21912.

## How was this patch tested?

Existing UTs

### CODE REVIEW

**Comment by kiszk** [comment]
retest this please

**Comment by kiszk** [comment]
retest this please

**Comment by kiszk** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 1]
```diff
@@ -3410,6 +3410,28 @@ case class ArrayDistinct(child: Expression)
     case _ => false
   }
 
+  @transient protected lazy val canUseSpecializedHashSet = elementType match {
```
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: Can we extract those common methods?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 2]
```diff
@@ -3517,56 +3510,24 @@ case class ArrayDistinct(child: Expression)
...
       primitiveValueTypeName: String): String = {
-    if (!isPrimitive) {
-      s"$distinctArray[$pos] = $getValue1";
-    } else {
-      s"$distinctArray.set$primitiveValueTypeName($pos, $getValue1)";
-    }
+    s"$distinctArray.set$primitiveValueTypeName($pos, $getValue1)"
   }
 
   private def setValueForFastEval(
```
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: Maybe we should rename this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala (hunk 1 above)
Comment (kiszk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: We can do so. To minimize the changes due to remaining time for cutting, I would like to do this in another PR #21912.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala [hunk 3]
```diff
@@ -3442,17 +3464,15 @@ case class ArrayDistinct(child: Expression)
...
-      if (elementTypeSupportEquals) {
+    if (canUseSpecializedHashSet) {
+      nullSafeCodeGen(ctx, ev, (array) => {
+        val i = ctx.freshName("i")
+        val sizeOfDistinctArray = ctx.freshName("sizeOfDistinctArray")
+        val foundNullElement = ctx.freshName("foundNullElement")
+        val openHashSet = classOf[OpenHashSet[_]].getName
+        val hs = ctx.freshName("hs")
+        val classTag = s"scala.reflect.ClassTag$$.MODULE$$.$hsTypeName()"
+        val getValue1 = CodeGenerator.getValue(array, elementType, i)
```
Comment (ueshin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala]: nit: `val getValue`?

**Comment by ueshin** [comment]
LGTM pending Jenkins.

**Comment by ueshin** [comment]
Thanks! merging to master.

**Pull request closed without merging**